raven-android/src/main/java/com/getsentry/raven/android/event/helper/AndroidEventBuilderHelper.java

```diff
@@ -46,6 +67,365 @@ public void helpBuildingEvent(EventBuilder eventBuilder) {
+     *
+     * @param ctx Android application context
+     * @return whether or not the device is currently plugged in and charging, or null if unknown
+     */
+    private static Boolean isCharging(Context ctx) {
+        try {
+            Intent intent = ctx.registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
+            if (intent == null) {
+                return null;
+            }
+
+            int plugged = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
+            return plugged == BatteryManager.BATTERY_PLUGGED_AC || plugged == BatteryManager.BATTERY_PLUGGED_USB;
+        } catch (Exception e) {
+            Log.e(TAG, "Error getting device charging state.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Get the device's current kernel version, as a string (from uname -a).
+     *
+     * @return the device's current kernel version, as a string (from uname -a)
+     */
+    private static String getKernelVersion() {
+        String errorMsg = "Exception while attempting to read kernel information";
+        BufferedReader br = null;
+        try {
+            Process p = Runtime.getRuntime().exec("uname -a");
+            if (p.waitFor() == 0) {
+                br = new BufferedReader(new InputStreamReader(p.getInputStream()));
+                return br.readLine();
+            }
+        } catch (Exception e) {
+            Log.e(TAG, errorMsg, e);
+        } finally {
+            if (br != null) {
+                try {
+                    br.close();
+                } catch (IOException ioe) {
+                    Log.e(TAG, errorMsg, ioe);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Attempt to discover if this device is currently rooted. From:
+     * https://stackoverflow.com/questions/1101380/determine-if-running-on-a-rooted-device
+     *
+     * @return true if heuristics show the device is probably rooted, otherwise false
+     */
+    private static Boolean isRooted() {
+        if (android.os.Build.TAGS != null && android.os.Build.TAGS.contains("test-keys")) {
+            return true;
+        }
+
+        String[] probableRootPaths = {
+            "/data/local/bin/su",
+            "/data/local/su",
+            "/data/local/xbin/su",
+            "/sbin/su",
+            "/su/bin",
+            "/su/bin/su",
+            "/system/app/SuperSU",
+            "/system/app/SuperSU.apk",
+            "/system/app/Superuser",
+            "/system/app/Superuser.apk",
+            "/system/bin/failsafe/su",
+            "/system/bin/su",
+            "/system/sd/xbin/su",
+            "/system/xbin/daemonsu",
+            "/system/xbin/su"
+        };
+
+        try {
+            for (String probableRootPath : probableRootPaths) {
+                if (new File(probableRootPath).exists()) {
+                    return true;
+                }
+            }
+        } catch (Exception e) {
```

Comment:
Should this catch be *inside* the loop to ensure the iterator is exhausted, rather than wrapping it? Was there anything specific you expected to catch here?